File: android/core/protocol/src/main/kotlin/com/optimus/voiceos/core/protocol/PhoneFraming.kt
package com.optimus.voiceos.core.protocol

import java.io.EOFException
import java.io.InputStream
import java.io.OutputStream

/** What a frame carries. Must match the PC's `PhoneFrameKind`. */
enum class PhoneFrameKind(val id: Byte) {
    JSON(1),
    AUDIO(2),
    TTS_AUDIO(3);

    companion object {
        fun fromId(id: Byte): PhoneFrameKind? = entries.firstOrNull { it.id == id }
    }
}

data class PhoneFrame(val kind: PhoneFrameKind, val payload: ByteArray) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is PhoneFrame) return false
        return kind == other.kind && payload.contentEquals(other.payload)
    }

    override fun hashCode(): Int = 31 * kind.hashCode() + payload.contentHashCode()
}

/**
 * The whole wire format: one byte of kind, four bytes of big-endian length, then the payload.
 *
 * Both ends ship together, so there is exactly one format and nothing to negotiate.
 */
object PhoneFraming {
    const val HEADER_BYTES = 5
    const val MAX_PAYLOAD_BYTES = 1024 * 1024

    fun encode(kind: PhoneFrameKind, payload: ByteArray): ByteArray {
        require(payload.size <= MAX_PAYLOAD_BYTES) {
            "Payload of ${payload.size} bytes exceeds the $MAX_PAYLOAD_BYTES limit."
        }

        val frame = ByteArray(HEADER_BYTES + payload.size)
        frame[0] = kind.id
        val length = payload.size
        frame[1] = (length ushr 24).toByte()
        frame[2] = (length ushr 16).toByte()
        frame[3] = (length ushr 8).toByte()
        frame[4] = length.toByte()
        payload.copyInto(frame, HEADER_BYTES)
        return frame
    }

    fun write(output: OutputStream, kind: PhoneFrameKind, payload: ByteArray) {
        output.write(encode(kind, payload))
        output.flush()
    }

    /** Reads exactly one frame, or returns null when the peer closed cleanly. */
    fun read(input: InputStream): PhoneFrame? {
        val header = ByteArray(HEADER_BYTES)
        if (!readFully(input, header)) return null

        val kind = PhoneFrameKind.fromId(header[0])
            ?: throw IllegalStateException("Unknown frame kind ${header[0]}.")

        val length = ((header[1].toInt() and 0xFF) shl 24) or
            ((header[2].toInt() and 0xFF) shl 16) or
            ((header[3].toInt() and 0xFF) shl 8) or
            (header[4].toInt() and 0xFF)

        if (length < 0 || length > MAX_PAYLOAD_BYTES) {
            throw IllegalStateException("Frame length $length is out of range.")
        }

        val payload = ByteArray(length)
        if (length > 0 && !readFully(input, payload)) throw EOFException("Truncated frame body.")

        return PhoneFrame(kind, payload)
    }

    private fun readFully(input: InputStream, buffer: ByteArray): Boolean {
        var offset = 0
        while (offset < buffer.size) {
            val read = input.read(buffer, offset, buffer.size - offset)
            if (read < 0) return false
            offset += read
        }
        return true
    }
}

File: src/Optimus.Core/Phone/PhoneFraming.cs
namespace Optimus.Core.Phone;

using System;
using System.Buffers.Binary;
using System.IO;
using System.Threading;
using System.Threading.Tasks;

/// <summary>What a frame carries.</summary>
public enum PhoneFrameKind : byte
{
    Json = 1,
    Audio = 2,
    TtsAudio = 3
}

/// <summary>
/// The whole wire format: one byte of kind, four bytes of big-endian length, then the payload.
/// </summary>
/// <remarks>
/// A raw socket with this header, rather than WebSocket or HTTP, because it needs no handshake,
/// no library on either side, and no version negotiation — all of which this slice rules out.
/// Both ends are shipped together, so there is exactly one format and nothing to negotiate.
/// </remarks>
public static class PhoneFraming
{
    public const int HeaderBytes = 5;

    /// <summary>Guards against a corrupt length turning into a huge allocation.</summary>
    public const int MaxPayloadBytes = 1024 * 1024;

    public static byte[] Encode(PhoneFrameKind kind, ReadOnlySpan<byte> payload)
    {
        if (payload.Length > MaxPayloadBytes)
        {
            throw new ArgumentException($"Payload of {payload.Length} bytes exceeds the {MaxPayloadBytes} limit.", nameof(payload));
        }

        byte[] frame = new byte[HeaderBytes + payload.Length];
        frame[0] = (byte)kind;
        BinaryPrimitives.WriteInt32BigEndian(frame.AsSpan(1, 4), payload.Length);
        payload.CopyTo(frame.AsSpan(HeaderBytes));
        return frame;
    }

    /// <summary>
    /// Reads exactly one frame. Returns null on a clean end of stream.
    /// </summary>
    public static async Task<(PhoneFrameKind Kind, byte[] Payload)?> ReadAsync(
        Stream stream,
        CancellationToken cancellationToken)
    {
        byte[] header = new byte[HeaderBytes];
        if (!await ReadExactlyAsync(stream, header, cancellationToken).ConfigureAwait(false))
        {
            return null;
        }

        var kind = (PhoneFrameKind)header[0];
        if (kind is not (PhoneFrameKind.Json or PhoneFrameKind.Audio or PhoneFrameKind.TtsAudio))
        {
            throw new InvalidDataException($"Unknown frame kind {header[0]}.");
        }

        int length = BinaryPrimitives.ReadInt32BigEndian(header.AsSpan(1, 4));
        if (length < 0 || length > MaxPayloadBytes)
        {
            throw new InvalidDataException($"Frame length {length} is out of range.");
        }

        byte[] payload = new byte[length];
        if (length > 0 && !await ReadExactlyAsync(stream, payload, cancellationToken).ConfigureAwait(false))
        {
            return null;
        }

        return (kind, payload);
    }

    private static async Task<bool> ReadExactlyAsync(Stream stream, byte[] buffer, CancellationToken cancellationToken)
    {
        int offset = 0;
        while (offset < buffer.Length)
        {
            int read = await stream
                .ReadAsync(buffer.AsMemory(offset, buffer.Length - offset), cancellationToken)
                .ConfigureAwait(false);

            if (read == 0)
            {
                return false; // Peer closed.
            }

            offset += read;
        }

        return true;
    }
}
